### PR TITLE
Feature/concurrent build fixes #67

### LINKED ISSUE #67: Race condition when multiple processes try to compile a module at once
Hi,

Great package by the way!

I've encountered an issue when multiple processes are spawned that all race to compile the same module. This can also occur when multiple processes are spawned on different hosts and share the same network filesystem. Such a situation is common when distributing work between multiple processes or hosts for AI or data analytics.

Here is a demonstration (in the shell):
```bash
echo '// cppimport
#include <pybind11/pybind11.h>

namespace py = pybind11;

int square(int x) {
    return x * x;
}

PYBIND11_MODULE(somecode, m) {
    m.def("square", &square);
}
/*
<%
setup_pybind11(cfg)
%>
*/' > somecode.cpp

echo 'import cppimport.import_hook
import somecode
somecode.square(9)' > test.py

rm somecode.cpython-*

for i in {1..100}; do python3 test.py & done
```

On my system around 4 out of 100 processes exit in an error. The shell output includes:
```
error: could not delete '/localdata/joshl/sandbox/somecode.cpython-36m-x86_64-linux-gnu.so': No such file or directory
...
Exit 1                  python3 test.py
...
Bus error               (core dumped) python3 test.py
```

These errors don't appear when the binary already exists.

----

To mitigate this issue in our applications we have used a file lock so that only one process attempts to compile the module at one time. A process first checks if the binary file exists, otherwise attempts to obtain the file lock. If it can't obtain the lock it waits until either the binary exists, can obtain the file lock or times out. Here is an example how it can be done ([app code](https://github.com/graphcore/examples-utils/blob/eba7f092c7f3c7d8875e562229540c7368eda1f0/examples_utils/load_lib_utils/cppimport_safe.py#L75)):

```python
from cppimport.checksum import is_checksum_valid

binary_path = module_data['ext_path']
lock_path = binary_path + '.lock'

t = time()

while not (os.path.exists(binary_path) and is_checksum_valid(module_data)) and time() - t < timeout:
    try:
        with FileLock(lock_path, timeout=1):
            if os.path.exists(binary_path) and is_checksum_valid(module_data_new_path):
                break
            # BUILD BINARY
            template_and_build(filepath, module_data)
    except Timeout:
        logging.debug(f'{os.getpid()}: Could not obtain lock')
        sleep(1)

if not (os.path.exists(binary_path) and is_checksum_valid(module_data_new_path)):
    raise Exception(
        f'Could not compile binary as lock already taken and timed out. Lock file will be deleted: {lock_path}')

if os.path.exists(lock_path):
    with suppress(OSError):
        os.remove(lock_path)
```

It would be great if we could upstream the above to `cppimport` to prevent the race condition errors. If you are happy with this solution I could contribute the above to the appropriate place in `cppimport`.

### PR DESCRIPTION
**Pull request by joshlk**
Fixes #67

### CODE REVIEW

File: cppimport/__init__.py [hunk 1]
```diff
@@ -60,16 +62,15 @@ def imp_from_filepath(filepath, fullname=None):
...
-        template_and_build,
-        try_load,
+        build_safely,
     )
-
+    filepath = os.path.abspath(filepath)
     if fullname is None:
         fullname = os.path.splitext(os.path.basename(filepath))[0]
     module_data = setup_module_data(fullname, filepath)
-    if is_build_needed(module_data) or not try_load(module_data):
```
[the remark this answers is not in the document]
Comment (tbenthompson) [reply cppimport/__init__.py:70]: Does the try_load logic cause problems in the multiprocess setting? I suspect there's a way to modify the logic so it maintains the old behavior while getting the multiprocess behavior to also be correct.

**Review by tbenthompson — APPROVED** [review]
This is awesome. Thank you so much! I have some comments through the PR about things that I'd like to do. I'm happy to resolve these if you're not interested or don't have time.

Also, I just added you as a collaborator on this project! That gives you permission to merge this in whenever you want - we'll have plenty of time to fix any small issues before another release.

Note:
- The actions workflow missed this PR because currently the workflow trigger is `on: [push]`. I need to change this to `on: [push, pull_request]` so that PRs created based on branches inside forked repos are caught by the trigger.

File: cppimport/__init__.py [hunk 2]
```diff
@@ -8,9 +8,11 @@
 from cppimport.find import _check_first_line_contains_cppimport
 
 settings = dict(
-    force_rebuild=False,
+    force_rebuild=False,  # `force_rebuild` with multiple processes is not supported
     file_exts=[".cpp", ".c"],
     rtld_flags=ctypes.RTLD_LOCAL,
+    lock_suffix='.lock',
+    lock_timeout=10*60,
```
[the remark this answers is not in the document]
Comment (joshlk) [reply cppimport/__init__.py:15]: Ive added a section in the README

File: cppimport/__init__.py [hunk 3]
```diff
@@ -8,9 +8,11 @@
 from cppimport.find import _check_first_line_contains_cppimport
 
 settings = dict(
-    force_rebuild=False,
+    force_rebuild=False,  # `force_rebuild` with multiple processes is not supported
```
[the remark this answers is not in the document]
Comment (joshlk) [reply cppimport/__init__.py:11]: Ive added a section in the README

**Comment by joshlk** [comment]
I tried modifying the CI yaml to include pull requests and it says:

> 1 workflow awaiting approval. First-time contributors need a maintainer to approve running workflows.

If you approve it - it should work

File: cppimport/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbenthompson) [reply cppimport/__init__.py:70]: I'll add a comment here to explain for the future. It's one of the more confusing bits of logic and the cppimport internals are not super well documented...

**Comment by tbenthompson** [comment]
Just a heads up that I'm going to fix the CI failures and make a few edits. Didn't want to duplicate the work with you. =)

**Comment by tbenthompson** [comment]
See the PR here: https://github.com/graphcore/cppimport-fork/pull/2

Copied the comment there:

> I opened this PR here so that it'll still be your PR that gets merged into cppimport.
> 
> pre-commit: black, isort, flake8 and corresponding small edits
> little comment about try_load
> raise an exception if a concurrent build is detected and force_rebuild is True
> changed "hook_test.cpp" to "tests/hook_test.cpp". The CI runs the tests from the project root so paths in the tests need to have that form.
> See the CI success here: https://github.com/tbenthompson/cppimport/pull/72 - I'll close that PR after yours is merged. I'm not worried about the codecov decrease. =/

The only remaining issue here is that one of the tests on Mac/Python 3.10 failed once: https://github.com/tbenthompson/cppimport/runs/7253199285?check_suite_focus=true

It's a stochastic failure and I can't replicate it. I can't quite explain this failure. My first guess was that it's happening when the built extension hasn't quite been fully written out. But that doesn't make sense because the file lock won't be released until after the extension has been fully written. So, given my lack of understanding I did two things:
1. I set `buffering=0` on the checksum trailer save so that it's saved immediately.
2. I added an OSError check on the checksum trailer load so that we'll just rebuild if something weird happens. This seems okay given the failure is rare.

The "Build and publish" CI step is failing because I set it up a bit wrong, but that shouldn't be your concern. 

I think this is 100% ready to merge. Let's do it!
1. You merge my PR into your branch
2. Then, we can merge this PR.

**Comment by tbenthompson** [comment]
Moved the two CI pipeline problems to their own issues: https://github.com/tbenthompson/cppimport/issues/74 https://github.com/tbenthompson/cppimport/issues/73

**Comment by joshlk** [comment]
Ok ready to merge 👍

**Pull request merged**